processor/src/main/java/org/mapstruct/ap/internal/util/SpecificCompilerWorkarounds.java

```diff
@@ -115,4 +135,161 @@ public static TypeElement replaceTypeElementIfNecessary(Elements elementUtils, T
         }
         return element;
     }
+
+    public static TypeMirror asMemberOf(DeclaredType containing, Element element, Types typeUtils,
+                                        Elements elementUtils) {
+        try {
+            // quick and dirty, until implementation below is more sound
+            return typeUtils.asMemberOf( containing, element );
+        }
+        catch ( IllegalArgumentException e ) {
+            TypeMirror eclipseWorkaround =
+                EclipseEnvironment.asMemberOf( containing, element, typeUtils, elementUtils );
+
+            if ( null != eclipseWorkaround ) {
+                return eclipseWorkaround;
+            }
+
+            throw e;
+        }
+    }
+
+    private static final class EclipseEnvironment {
+        public static TypeMirror asMemberOf(DeclaredType containing, Element element, Types typeUtils,
+                                            Elements elementUtils) {
+            if ( typeUtils instanceof TypesImpl && containing instanceof DeclaredTypeImpl
```

Comment:
Just a tiny style/test thing, but I thought I'd bring it up: I'd prefer to invert the condition and return null right here. That way all the remainder of the method can be indented one level less which I think is good for readability. Your call to change it or not :)